### DISCUSSION TITLE
How to set SLURM gres parameter

### DISCUSSION THREAD

**Discussion by JuhaHuiskonen** [nextpyp/.github#43] [Q&A] [answered]
On our HPC environment, the number of GPUs must be reserved with `gres` parameter.  For example: `#SBATCH --gres=gpu:v100:1` reservers one GPU.  I understand currently setting additional SLURM parameters is not possible in the `config.toml`.  If anyone knows a workaround, please let me know!

**Comment by abartesaghi** [comment]
The new version [0.6.1](https://nextpyp.app/files/pyp/0.6.1/docs/index.html) allows you to specify GPU resources using the Gres parameters in the **Resources** tab. In your case, you will need to set `Gres per task` to `gpu:v100:1`

**Comment by JuhaHuiskonen** [comment]
How is this linked to a specific queue?   For instance,  I would like to reserve both fast local disk and GPUs (`nvme:3600,gpu:v100:4`) when I select queue `gpu`, but only the fast disk (`nvm:3600`)  when I select queue `cpu`. As these configurations are complex, it might be easier to provide a SLURM batch job template file for each different queue type. It would also be useful to see the SLURM batch script that is created for debugging purposes. Now a small mistake in SLURM parameters seems to give `Internal server error` but no other information (if the SLURM job fails on submission).

Edit: looks like the gres field cannot be empty (even when submitting to CPU queue and where no "gres resources" are sometimes not needed).  If I give `nvme:0` (so no fast scratch disk reserved) instead of just leaving this empty, the job seems to start. 

Edit2:` local/logs/micromon` is the place to check for possible errors occurring in SLURM job submission. It would be great if these errors would be reported in the Web GUI.

**Reply by abartesaghi** [reply]
- When you first configure nextPYP, you need to specify a CPU and a GPU queue or partition(s). If a job needs a GPU to run, then nextPYP will automatically send these jobs to the GPU partition and all others to the CPU one. The gres options allow you to override this behavior, if you specify a gpu resource, you will be forcing those jobs to be submitted to a GPU node even if the job doesn't need access to a GPU.
- Individual nextPYP jobs do not use multiple GPUs, so you will always want to set gres to use a single gpu, e.g.,: `gpu:1`. In your case, you will use: `nvme:3600,gpu:v100:1` in the gres options. 
- You should be able to use an empty gres field, that's actually the default value. We will do some testing to see if we can reproduce this issue.
- Indeed, the micromon log is where these errors will appear. Thanks for the feedback & suggestions!

**Reply by JuhaHuiskonen** [reply]
My config.toml has this 

```
[slurm]
# configure SLURM partitions here
queues = [ "small", "large", "gpu", "gputest" ]
# configure GPU partitions here
gpuQueues = [ "gpu", "gputest" ]

```

The resources section of the run B2 in tomography tutorial (preprocessing) has these options:

<img width="1506" alt="Screenshot 2024-01-31 at 20 13 06" src="https://github.com/nextpyp/.github/assets/20352352/7e6bcd8d-7141-4b21-829c-a98f00b6e440">

<img width="1489" alt="Screenshot 2024-01-31 at 20 13 19" src="https://github.com/nextpyp/.github/assets/20352352/2a9e1883-a6b3-403b-9604-dc12fb14ad60">

The "launch" part of the job gets submitted correctly, but then fails with this error message:

```
1	2024-01-31 20:11:19 [INFO] /opt/pyp/bin/run/pyp:3273 | Job 20281823 (v0.6.1) launching on r18c47.bullx using 1 task(s) and 4 GB of RAM
2	2024-01-31 20:11:19 [INFO] /opt/pyp/bin/run/pyp:4645 | Running on directory /scratch/project_2007321/users/huiskone/nextPYP/shared/users/huiskone/projects/tomo-tutorial-GPU-EdBoIjH4XOYp5gV0lyT4rEj/tomo-preprocessing-ewcoTCdDqNtm1cK8
3	2024-01-31 20:11:19 [INFO] pyp/system/project_params.py:763 | Parameters changed: {'slurm_merge_memory', 'slurm_queue', 'slurm_merge_queue', 'slurm_walltime', 'slurm_max_cpus', 'slurm_merge_gres', 'slurm_merge_tasks'}
4	2024-01-31 20:11:19 [INFO] pyp/system/project_params.py:164 | log/TS_01.log does not exist
5	2024-01-31 20:11:19 [INFO] pyp/system/project_params.py:164 | log/TS_03.log does not exist
6	2024-01-31 20:11:19 [INFO] pyp/system/project_params.py:164 | log/TS_43.log does not exist
7	2024-01-31 20:11:19 [INFO] pyp/system/project_params.py:164 | log/TS_45.log does not exist
8	2024-01-31 20:11:19 [INFO] pyp/system/project_params.py:164 | log/TS_54.log does not exist
9	2024-01-31 20:11:19 [INFO] pyp/system/db_comm.py:35 | Parameters entered into database successfully
10	Traceback (most recent call last):
11	  File "/opt/pyp/bin/run/pyp", line 4660, in <module>
12	    split(parameters)
13	  File "/opt/pyp/bin/run/pyp", line 951, in split
14	    id = slurm.submit_jobs(
15	  File "/opt/pyp/src/pyp/system/slurm.py", line 409, in submit_jobs
16	    id = jobs.submit_commands(
17	  File "/opt/pyp/src/pyp/streampyp/jobs.py", line 226, in submit_commands
18	    return Web().slurm_sbatch(
19	  File "/opt/pyp/src/pyp/streampyp/web.py", line 136, in slurm_sbatch
20	    return self._request(
21	  File "/opt/pyp/src/pyp/streampyp/web.py", line 39, in _request
22	    response = self.client.send(request)
23	  File "/usr/local/envs/pyp/lib/python3.8/site-packages/apply_defaults/decorators.py", line 13, in wrapper
24	    return function(self, *args, **kwargs)
25	  File "/usr/local/envs/pyp/lib/python3.8/site-packages/jsonrpcclient/client.py", line 177, in send
26	    raise ReceivedErrorResponseError(response.data)
27	jsonrpcclient.exceptions.ReceivedErrorResponseError: Cluster job submission failed:
28	unable to read job id from sbatch output:
29	sbatch: error: Batch job submission failed: Invalid generic resource (gres) specification
30	2024-01-31 20:11:20 [INFO] /opt/pyp/bin/run/pyp:3200 | Filesystem                                      Size  Used Avail Use% Mounted on
31	2024-01-31 20:11:20 [INFO] /opt/pyp/bin/run/pyp:3200 | 10.140.15.146@o2ib:10.140.15.145@o2ib:/scratch   20T  3.7T   16T  19% /scratch/project_2007321/users/huiskone/nextPYP/tmp
32	2024-01-31 20:11:20 [INFO] /opt/pyp/bin/run/pyp:4686 | Deleted temporary files from /scratch/project_2007321/users/huiskone/nextPYP/tmp/huiskone/20281823
33	
```

If nextPYP would show the SLURM submission command that could help to debug this.

**Reply by abartesaghi** [reply]
This is the relevant part of the error:
`sbatch: error: Batch job submission failed: Invalid generic resource (gres) specification`
With your settings, you will be sending jobs to the `large` partition, with 1 cpu, 16GB of RAM, a v100 gpu, 0 nvme (not sure exactly what that means), and a walltime of 3 days. Is this a valid combination in your cluster? If you selected motioncor3 or aretomo in the pre-processing options, then the jobs will be sent to the `gpu` partition, so you'll need to make sure that the above is a valid combination for that partition. We will definitely look into exposing the actual sbatch command for debugging purposes.

**Reply by JuhaHuiskonen** [reply]
I am attempting a GPU run and selected AreTomo for tilt series alignment and reconstruction.  So I am assuming the job would be sent to partition `gpu` and not `large`.   If it is sent to `large`, then I would understand the error (this partition does not have GPUs).  However, if it is correctly sent to `gpu` then I don't understand the error as the parameters should be correct.  One GPU node has 20 CPUs, 4 GPUs, and plenty of RAM (nvme refers to the amount of fast local disk on the node, that was needed here as the value was 0)

edit: are GPU jobs submitted as array jobs?  Maybe this is not allowed on my system and causes the gres error.

**Reply by abartesaghi** [reply]
If you selected AreTomo for alignment and reconstruction, then the jobs will be submitted to the `gpu` queue. Yes, all jobs are submitted as job arrays. If you set the bundle size to match the number of tilt-series, then the job array will have a single component but will still be submitted as a job array of size 1. It could be worth trying that option to get more information.

**Reply by JuhaHuiskonen** [reply]
I run with `Threads per task= 1` and `Max number of threads=1` for project where I had imported only one tilt-series. This created 1 sub-job.  The error is the same.

**Reply by abartesaghi** [reply, marked as answer]
We just realized that the web server was not interpreting the gres options correctly. We will fix this and push a new patch release.

**Reply by abartesaghi** [reply]
These issues are addressed in version [0.6.2](https://nextpyp.app/files/pyp/0.6.2/docs/index.html).

**Comment by jpellman** [comment]
Just as a general suggestion, instead of having multiple parameters in the `[slurm]` section of the TOML config map to `sbatch` directives (e.g., `memoryPerNode`, `cpusPerNode`,`queues`,`gpuqueues`) or setting job parameters into a hardcoded template via the UI it might make sense to allow users and admins to provide a batch job template file.  In this way, you could give users the ability to inject arbitrary values into the job template and you wouldn't need to worry about adding every possible job option to the TOML config (or the web UI).  This would allow job configuration to be more open-ended, and would also make it much easier to support job schedulers other than SLURM (e.g., LSF, PBS, Kubernetes).  

You could even use an already existing templating engine (e.g., Go templates, `jinja2`) and not have to worry about implementing your own templating logic (i.e., more code to maintain).

This is what RELION and CryoSPARC do and it works relatively well for our purposes.

**Reply by cuchaz** [reply]
Hi there. We're working on expanding the customizability of our SLURM options and wanted to know more about the use cases you had in mind. Unfortunately for us, supporting other job schedulers isn't as simple as exposing a few options in a template, but there's still a lot we can do to make the SLURM option system more flexible.

We're very wary of just opening up all the SLURM variables in a template because that would make it far too easy for someone to break the website functionality by, for example, overriding a SLURM option that was intended to be a very specific value. Plus, many of the SLURM options already have a UI that allows people to configure them for each job submission.

However, there's no harm in allowing additional SLURM options that wouldn't conflict with our code. Would the ability to set additional options in an ad hoc way in the config file satisfy your use cases? Or maybe even in the SLURM resources UI? What would you like to be able to do specifically that isn't available in the current system?

**Reply by Qiangong2** [reply]
I think what @jpellman is trying to say is that if job template support is added, then the nextPYP team won't have to manually implement each variable. Most institutions use some form of lua job submit plugins with their SLURM instance to regulate which variables are required to submit a job. Having nextPYP read a template allows an institution to utilize their custom variables without needing to manually edit the source code.

**Reply by cuchaz** [reply]
I'm not suggesting we manually implement every variable. I'm suggesting we build a system that lets people use arbitrary variables just like templates would, but the system would be based on something other than templating, because templating has downsides for us. I'd just like to know what people's specific use cases are, so I can get an idea if our approach will work or not.

**Reply by Qiangong2** [reply]
I guess I'm confused what the downsides would be with templating? As long as you've set up your scheduler correctly, there should be no issues.

What we do with Relion and CryoSPARC is that we have a base template that maps the variables that Relion and CryoSPARC provide, which then submits it to the SLURM cluster. Relion allows us to insert extra variables into the running tab and CryoSPARC allows you to set pretty much anything. Something like that would be ideal for nextPYP. As it is, it's a major blocker for using it in our environment.

**Reply by cuchaz** [reply]
Ok, so it sounds like you're mostly interested in inserting extra variables into the SLURM job? Are these extra variable values essentially static? ie, they're the same for every user and every job in every situation? Or do they vary depending on context: like the current user, the job type, etc

**Reply by jpellman** [reply]
@cuchaz 

I'm encountering no particular problem currently / have no particular use case that I can describe.  My suggestion was based on my experience administering SLURM clusters and my general understanding regarding application development best practices.

> We're very wary of just opening up all the SLURM variables in a template because that would make it far too easy for someone to break the website functionality by, for example, overriding a SLURM option that was intended to be a very specific value.

It sounds like what you're saying is that you want nextPYP to validate job state prior to job submission, with the expectation that job state will remain static after submission.  Generally speaking this is a dangerous assumption to make when interfacing an application with a system-level component (such as an OS scheduler, multicomputer batch scheduler or orchestrator).  

In SLURM, for example, there are myriad ways to alter what resources a job requests and other job parameters after submission.  @Qiangong2 has mentioned one such mechanism ([lua job submit plugins](https://slurm.schedmd.com/job_submit_plugins.html), which act as mutators/hooks that run upon job submission).  Additionally, job specifications can be altered using environment variables (e.g., `SBATCH_GRES`, which overrides the `--gres` directive defined within a batch script) or via the `scontrol` command (via `scontrol update JobId=<ID> ...`).  SLURM exposes many many mechanisms that change/update job state, and writing code around all of these mechanisms will likely be a messy affair.

As an alternative, it might make sense for your web application to maintain its own view of the job state and then periodically poll the scheduler to see if there is drift between the application's view of state and SLURM's view of state.  In the event of an inconsistency, the web application could then take a corrective action (e.g., running `scontrol update`, cancelling the job and displaying an error).  This might ultimately be a more robust solution than trying to validate that a batch job will be compatible with the web application at job submission time and assuming that job state will remain static.

> Would the ability to set additional options in an ad hoc way in the config file satisfy your use cases? Or maybe even in the SLURM resources UI?

The fundamental issue I foresee with such a setup would be that this sort of design presupposes that the application is more context-aware than the end-users and system administrators running the application.  There are many, many idiosyncratic batch cluster configurations out there, and trying to cover every single edge case within application logic creates a lot of work.  By letting end-users/system administrators use a template, you give individuals with more awareness of their environments the ability to deal with those edge cases instead of adding extra code to your program.

> What would you like to be able to do specifically that isn't available in the current system?

I have no particular complaints thus far, only observations and suggestions that could potentially prevent future complaints from arising.  Ultimately this is your application and you have a better vision of what you want to accomplish with it.  You are welcome to adopt, discard or critique my commentary as necessary for your purposes.

**Reply by asarnow** [reply]
@cuchaz I would specifically like to be able to blacklist (or whitelist) a few specific nodes.

I would also like to be sure that the correct GPU is being used so that multiple 1 GPU jobs can run on one node, but I assume nextPYP is reading the SLURM environment variable set by the gres argument so this is fine.

**Reply by cuchaz** [reply]
@jpellman Hi there, and thanks for the super helpful feedback!

I totally agree that once a job goes out to the scheduler, anything can happen to the submission settings and our application shouldn't try to control what's going on downstream. Luckily for us, our concerns are much simpler.

My earlier comment about ensuring certain SLURM variables make it to the job scheduler without changes ended up being somewhat wrong. While we do expose some basic SLURM options in the UI that can be set for each job, if something intercepts those values and changes them, it may confuse the user, but it won't break anything in our application. We'd like to avoid user confusion where we can, but as you mentioned, SLURM plugins may subvert expectations regardless and they're completely out of our control anyway.

Our main objection to using templates is mostly technical. Currently we configure SLURM job submissions using the sbatch command line arguments. Since we support job schedulers other than SLURM (like our standalone mode), none of the scheduler-specific configuration is included in the submitted batch script itself. So this setup isn't really amenable to templating. We could switch to the comment-style method of configuring jobs inside the batch script (which would be more amenable to templating), but it would be a bit of work for us. I was hoping there was a simpler way to give people the flexibility they want, but unfortunately I'd need more information about actual use cases to come up with anything reasonable here.

**Reply by cuchaz** [reply]
@asarnow Thanks for the feedback. This is really helpful. A few follow-up questions if you don't mind.

Is the kind of node filtering you suggested supported by any of the other Cryo software packages? Is that something their templating systems allow you to do? What would such a template look like?

In our deployments, we handle node filtering using the SLURM partitions (though the nextPYP UI calls them queues). Is that approach available to you, or do you need this feature in nextPYP itself?

As for the GPUs, I think we recently improved GPU resource handling, so multiple jobs landing on the same node should end up using the GPUs assigned to them by SLURM.

**Reply by asarnow** [reply]
In cryoSPARC a cluster template is imported as a "lane" that can be chosen for jobs, the template can contain arbitrary SLURM parameters such as a node blacklist. In Relion, a template script file is selected through the GUI, so same. These scripts also allowed us to manually choose the right GPUs for the software, in case the queuing system didn't manage it on its own.

We have 1 or 2 nodes in our gpu partition that have some flaky PCI slots, blacklisting them is an immediate workaround without all the chores / time we should do for a permanent fix. (Also don't want to just shut down the nodes at this time, they are doing some work fine).

**Reply by cuchaz** [reply]
Ah, I see. That makes sense. Thanks for the info! I'll admit I'm not an expert on how Relion and cryoSPARC work. =P

It sounds like being able to pick the right template for each job is a really important tool.
